### PR TITLE
Penumbra support v2

### PR DESCRIPTION
**Pull request by hdevalence**
Replaces #1124 but from the Penumbra fork to make collaboration easier

### CODE REVIEW

**Comment by conorsch** [comment]
Using the latest code in this branch, along with the configs in https://github.com/penumbra-zone/penumbra/tree/fa2f750f4ec049bc0d7195dc368fc7490f5c347e/deployments/relayer , we're able to build a working path:

```
❯ rly paths list
 0: penumbra_path        -> chns(✔) clnts(✔) conn(✔) (penumbra-testnet-pasiphae<>penumbra-testnet-pasiphae-5d99ea66)
 ```

File: relayer/client.go [hunk 1]
```diff
@@ -510,6 +510,7 @@ func findMatchingClient(ctx context.Context, src, dst *Chain, newClientState ibc
 // parseClientIDFromEvents parses events emitted from a MsgCreateClient and returns the
 // client identifier.
 func parseClientIDFromEvents(events []provider.RelayerEvent) (string, error) {
+	fmt.Printf("client.go events: %+v\n", events)
```
Comment (jtieri) [inline relayer/client.go]: looks like an artifact from debugging, this could be removed

File: relayer/channel.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func (c *Chain) CreateOpenChannels(
 			dst.chainProcessor(c.log, nil),
 		).
 		WithPathProcessors(pp).
-		WithInitialBlockHistory(0).
+		WithInitialBlockHistory(100).
```
Comment (jtieri) [inline relayer/channel.go:77]: is there a specific reason for this change?

File: relayer/events.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			}
+		}
+	}
+	return "", fmt.Errorf("client identifier event attribute not found")
+}
+
+// ParseConnectionIDFromEvents parses events emitted from a MsgConnectionOpenInit or
+// MsgConnectionOpenTry and returns the connection identifier.
+func ParseConnectionIDFromEvents(events []provider.RelayerEvent) (string, error) {
+	fmt.Printf("events.go events: %+v\n", events)
```
Comment (jtieri) [inline relayer/events.go]: another print statement that could go

File: relayer/chains/penumbra/penumbra_chain_processor.go [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+	"github.com/avast/retry-go/v4"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	ctypes "github.com/cometbft/cometbft/rpc/core/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	conntypes "github.com/cosmos/ibc-go/v7/modules/core/03-connection/types"
+	chantypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	"github.com/cosmos/relayer/v2/relayer/processor"
+	"github.com/cosmos/relayer/v2/relayer/provider"
+
+	//	jsonrpcclient "github.com/cometbft/cometbft/rpc/jsonrpc/client"
```
Comment (jtieri) [inline relayer/chains/penumbra/penumbra_chain_processor.go]: commented import statement could be removed

File: relayer/chains/penumbra/penumbra_chain_processor.go [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+		})
+	}
+
+	persistence.latestQueriedBlock = newLatestQueriedBlock
+
+	return nil
+}
+
+/*
+func (pcp *PenumbraChainProcessor) queryCycle(ctx context.Context, persistence *queryCyclePersistence) error {
```
Comment (jtieri) [inline relayer/chains/penumbra/penumbra_chain_processor.go]: if this is no longer being used as a reference we could pull this out

File: relayer/chains/penumbra/tx.go [hunk 6]
```diff
@@ -0,0 +1,2289 @@
...
+		return &penumbracrypto.MerkleRoot{Inner: res.Value[2:]}, nil
+	}
+
+	return &penumbracrypto.MerkleRoot{Inner: res.Value[2:]}, nil
+}
+
+func parseEventsFromABCIResponse(resp abci.ResponseDeliverTx) []provider.RelayerEvent {
+	var events []provider.RelayerEvent
+
+	fmt.Printf("tx.go events: %+v\n", resp.Events)
```
Comment (jtieri) [inline relayer/chains/penumbra/tx.go]: another print statement that could be removed

File: relayer/chains/penumbra/tx.go [hunk 7]
```diff
@@ -0,0 +1,2289 @@
...
+				continue
+			}
+			attributes[string(key)] = string(value)
+		}
+		events = append(events, provider.RelayerEvent{
+			EventType:  event.Type,
+			Attributes: attributes,
+		})
+	}
+	fmt.Printf("parsed events: %+v\n", events)
```
Comment (jtieri) [inline relayer/chains/penumbra/tx.go]: here as well

File: relayer/chains/penumbra/tx.go [hunk 8]
```diff
@@ -0,0 +1,2289 @@
...
+	var height int64
+	var data []byte
+	var txhash string
+	var code uint32
+
+	syncRes, err := cc.sendMessagesInner(ctx, msgs, _memo)
+	if err != nil {
+		return nil, false, err
+	}
+	cc.log.Info("waiting for penumbra tx to commit", zap.String("syncRes", fmt.Sprintf("%+v", syncRes)))
```
Comment (jtieri) [inline relayer/chains/penumbra/tx.go]: can we capitalize the first word in this log statement? for reference we do have a style guide for logging which can be found [here](https://github.com/cosmos/relayer/blob/main/docs/logging.md)

File: relayer/chains/penumbra/tx.go [hunk 9]
```diff
@@ -0,0 +1,2289 @@
...
+
+	if err := retry.Do(func() error {
+		ctx, cancel := context.WithTimeout(ctx, 40*time.Second)
+		defer cancel()
+
+		res, err := cc.RPCClient.Tx(ctx, syncRes.Hash, false)
+		if err != nil {
+			return err
+		}
+		cc.log.Info("got penumbra tx result", zap.String("res", fmt.Sprintf("%+v", res)))
```
Comment (jtieri) [inline relayer/chains/penumbra/tx.go]: See above comment

File: relayer/chains/penumbra/tx.go [hunk 10]
```diff
@@ -0,0 +1,2289 @@
...
+	}
+
+	return clientState, nil
+}
+
+// DefaultUpgradePath is the default IBC upgrade path set for an on-chain light client
+var defaultUpgradePath = []string{"upgrade", "upgradedIBCState"}
+
+/*
+fn apphash_spec() -> ics23::ProofSpec {
```
Comment (jtieri) [inline relayer/chains/penumbra/tx.go]: this looks like it was being used to port some Rust code but if it's no longer needed it could be removed

**Review by jtieri — COMMENTED** [review]
great work penumbra team 🔥 🔥 
just a handful of nits but otherwise this lgtm

**Review by jtieri — APPROVED** [review]

**Comment by conorsch** [comment]
Thanks for pointing me to the style guide, @jtieri. I've updated all the places you called out, and touched up a few more logging statements to be compliant. After making these changes, I rebuilt and reran the path build script, to ensure I didn't inadventerntly break something. Indeed, we're still good:

```
❯ rly paths list
 0: penumbra_path        -> chns(✔) clnts(✔) conn(✔) (penumbra-testnet-pasiphae<>penumbra-testnet-pasiphae-7549032f)
 ```

**Review by jtieri — APPROVED** [review]
🚢 🚢

**Pull request merged**